text/0000-futures.md

```diff
@@ -154,31 +164,323 @@ can be dropped.
+    pub data: *const (),
+    /// Virtual function pointer table that customizes the behavior of this waker.
+    pub vtable: &'static RawWakerVTable,
+}
+
+/// A virtual function pointer table (vtable) that allows to customize the
+/// behavior of a `RawWaker`.
+/// The pointers which get passed to all functions inside the vtable are the
+/// values of the `data` field.
+#[derive(PartialEq, Copy, Clone)]
+pub struct RawWakerVTable {
+    /// This function will be called when the `RawWaker` gets cloned, e.g. when
+    /// the `Waker` or `LocalWaker` in which the `RawWaker` is stored gets cloned.
+    ///
+    /// The implementation of this function must retain all resources that are
+    /// required for this additional instance of a `RawWaker` and associated
+    /// task. The implementation must return a valid `RawWaker` that behaves
+    /// equivalent to the `RawWaker` that got cloned. E.g. cloning a `RawWaker`
+    /// that implemented a thread-safe wakeup for use in `Waker` must return
+    /// a `RawWaker` that implements the same wakeup behavior.
+    pub clone: unsafe fn(*const ()) -> RawWaker,
+    /// This function will be called when a `LocalWaker` should be converted into
+    /// a thread-safe `Waker`. The implementation of this function must return
+    /// a new `RawWaker` which can fulfill the requirements of `Waker`. It can
+    /// exchange the vtable for that purpose. E.g. it might replace the `wake`
+    /// function with a varient that supports cross-thread wakeups.
+    ///
+    /// The old `LocalWaker` which contained the data pointer should be consumed
+    /// while performing the operation. After the operation the `LocalWaker`
+    /// won't exist anymore, only the new `Waker`.
+    /// This means that if both instances would utilize the same reference-counted
+    /// data, changing the reference count would not be necessary.
+    ///
+    /// If a conversion is not supported, the implementation should return
+    /// `None`. In this case the implementation must still make
+    /// sure that the data pointer which is passed to the function is correctly
+    /// released, e.g. by calling the associated `drop_fn` on it.
+    pub into_waker: unsafe fn(*const ()) -> Option<RawWaker>,
+    /// This function will be called when `wake` is called on the `RawWaker`.
+    pub wake: unsafe fn(*const ()),
+    /// This function gets called when a `RawWaker` gets dropped.
+    ///
+    /// The implementation of this function must make sure to release any
+    /// resources that are associated with this instance of a `RawWaker` and
+    /// associated task.
+    pub drop_fn: unsafe fn(*const ()),
+}
+
+/// A `Waker` is a handle for waking up a task by notifying its executor that it
+/// is ready to be run.
+///
+/// This handle encapsulates a `RawWaker` instance, which defines the
+/// executor-specific wakeup behavior.
+///
+/// Implements `Clone`, `Send`, and `Sync`.
+pub struct Waker {
+    waker: RawWaker,
+}
+
```

Comment:
Although it is clearly stated in the documentation, it would still probably be good to include the core difference between `Waker` and `LocalWaker` (the `Send` and `Sync` implementations) in the code snippet.

```suggestion

+unsafe impl Send for Waker { }
+unsafe impl Sync for Waker { }

```